src/v/raft/mux_state_machine.h

```diff
@@ -214,31 +209,69 @@ ss::future<std::error_code> mux_state_machine<T...>::replicate_and_wait(
   model::record_batch&& b,
   model::timeout_clock::time_point timeout,
   ss::abort_source& as) {
-    using ret_t = std::error_code;
-    return ss::with_gate(
-      _gate, [this, b = std::move(b), timeout, &as]() mutable {
-          return _mutex.get_units().then([this, b = std::move(b), timeout, &as](
-                                           ss::semaphore_units<> u) mutable {
-              return replicate(std::move(b))
-                .then([this, u = std::move(u), timeout, &as](
-                        result<raft::replicate_result> r) {
-                    if (!r) {
-                        return ss::make_ready_future<ret_t>(r.error());
-                    }
-
-                    auto last_offset = r.value().last_offset;
-
-                    auto [it, insterted] = _promises.emplace(
-                      last_offset, expiring_promise<ret_t>{});
-                    vassert(
-                      insterted,
-                      "Prosmise for offset {} already registered",
-                      last_offset);
-                    return it->second.get_future_with_timeout(
-                      timeout, [] { return errc::timeout; }, as);
-                });
-          });
+    if (_gate.is_closed()) {
+        return ss::make_ready_future<std::error_code>(errc::shutting_down);
+    }
+
+    auto u = get_units();
+    auto promise = std::make_unique<promise_t>();
+    auto f = promise->get_future_with_timeout(
+      timeout, [] { return make_error_code(errc::timeout); }, as);
+
+    ssx::spawn_with_gate(
+      _gate,
+      [this,
+       b = std::move(b),
+       u = std::move(u),
+       promise = std::move(promise)]() mutable {
+          return replicate(std::move(b))
+            .then_wrapped(
+              [this, u = std::move(u), promise = std::move(promise)](
+                ss::future<result<raft::replicate_result>> f) mutable {
+                  if (f.failed()) {
+                      promise->set_exception(f.get_exception());
+                      return ss::now();
+                  }
+                  auto r = f.get();
+                  if (!r) {
+                      promise->set_value(r.error());
+                      return ss::now();
+                  }
+
+                  auto last_offset = r.value().last_offset;
+
+                  return _new_result
+                    .wait([this, last_offset] {
+                        return _last_applied >= last_offset;
+                    })
+                    .then_wrapped(
+                      [this,
+                       last_offset,
+                       u = std::move(u),
+                       promise = std::move(promise)](ss::future<> f) mutable {
+                          u.release();
+                          if (f.failed()) {
+                              promise->set_exception(f.get_exception());
+                              return;
+                          }
```

Comment:
if u.release() cases the pending count to go to zero, but here we return would the next time apply runs it would clear the results even though there may have been an ephemeral failure? the concurrency control here is sort of hard to understand. do you think that is a concerning scenario?